[New] Real-time web socket tutorial rewrite (#2408)
timescale/docs: tutorials/ingest-real-time-websocket-data.md

Before you begin, make sure you have:

*   A Timescale [service and connect to the service][financial-tick-dataset].

> *   Download the `.sql` file that contains the credentials such as `<HOST>`,

Reviewer: ```suggestion
*   Downloaded the file that contains your Timescale service credentials such as `<HOST>`,
```

>     `<PORT>`, `<PASSWORD>`, and other details to connect to your Timescale
>     service. You can also find these details in the `Connection Info` section of
>     your Timescale service in the Timescale portal.

Reviewer: ```suggestion
    `<PORT>`, and `<PASSWORD>`. Alternatively, you can find these details 
    in the `Connection Info` section for your service.
```

*   Installed Python 3
*   Signed up for [Twelve Data][twelve-signup]. The free tier is perfect for
    this tutorial.
*   Made a note of your Twelve Data [API key](https://twelvedata.com/account/api-keys).

<Collapsible heading="Connect to the websocket server">

When you connect to the Twelve Data API through a websocket, you create a
persistent connection between your computer and the websocket server. 
You set up a Python environment, and pass two arguments to create a
websocket object and establish the connection.

> ## Set up a new Python environment

Reviewer: I can't suggest it, but L67 in this file needs "easy" removed. Something like "This library allows you to make requests ..."

    This argument needs to be a list of stock ticker symbols (for example,
    `MSFT`) or crypto trading pairs (for example, `BTC/USD`). When using a
    websocket connection you always need to subscribe to the events you want to
    receive. You can do this by using the `symbols` argument or if your
    connection is already created you can also use the `subscribe()` function to
    get data for additional symbols.

<Procedure>

### Connecting to the websocket server

1.  Create a new Python file called `websocket_test.py` and connect to the
    Twelve Data servers using the `<YOUR_API_KEY>`:

    ```python
       import time
       from twelvedata import TDClient

        messages_history = []

        def on_event(event):
         print(event) # prints out the data record (dictionary)
         messages_history.append(event)

       td = TDClient(apikey="<YOUR_API_KEY>")
       ws = td.websocket(symbols=["BTC/USD", "ETH/USD"], on_event=on_event)
       ws.subscribe(['ETH/BTC', 'AAPL'])
       ws.connect()
       while True:
       print('messages received: ', len(messages_history))
       ws.heartbeat()
       time.sleep(10)
    ```

1.  Run the Python script:

    ```bash

>     python websocket_test.py

Reviewer: I can't suggest it, L146 is a result statement, so it should be in the same step as the previous. I'm also a bit 🫤 on the "immediately", maybe say "When you run the script, you receive a response from the server ..." ? 

Also, it's not highlighting properly. Could be a whitespace problem?

![image](https://github.com/timescale/docs/assets/3914967/ef8c63b8-6f9c-4883-994f-7df94d4750b4)

     'fails': None
    }
    ```
When you have established a connection to the websocket server,
wait a few seconds, and you can see data records, like this:

> ```bash

Reviewer: This code block isn't indented properly:

![image](https://github.com/timescale/docs/assets/3914967/3c42f0bc-c53f-43c9-a00d-a5bcdd973982)

such as the name of the exchange, and the current price. You can also
occasionally see `heartbeat` events in the response; these events signal
the health of the connection over time.
At this point the websocket connection is working successfully to pass data.
    
</Procedure>
    
To ingest the data into your Timescale service, you need to implement the 
`on_event` function.    
</Collapsible>

> <Collapsible heading="The real-time dataset">

Reviewer: Collapsibles are automatically L3 headings, which means you have some weirdness going on with heading levels here:

![image](https://github.com/timescale/docs/assets/3914967/cf0a5684-6b3f-4325-95e0-d0207995d4c8)

Using batching is a fairly common pattern when ingesting data into TimescaleDB
from Kafka, Kinesis, or websocket connections.

You can implement a batching solution in Python with Psycopg2.
You can implement the ingestion logic within the `on_event` function that
you can then pass over to the websocket object.

This function needs to:

1.  Check if the item is a data item, and not websocket metadata.
1.  Adjust the data so that it fits the database schema, including the data
    types, and order of columns.
1.  Add it to the in-memory batch, which is a list in Python.
1.  If the batch reaches a certain size, insert the data, and reset or empty the list.

## Ingesting data in real-time

<Procedure>

1.  Update the Python script that prints out the current batch size, so you can
    follow when data gets ingested from memory into your database. Use
    the `<HOST>`, `<PASSWORD>`, and `<PORT>` details for the Timescale service
    where you want to ingest the data and your API key from Twelve Data:

> ```python

Reviewer: Both of the code blocks in this section aren't indented properly, so it's resetting the second step to be step 1:

![image](https://github.com/timescale/docs/assets/3914967/c7678413-a6d8-4066-b847-a2e09edb33ae)

Using batching is a fairly common pattern when ingesting data into TimescaleDB
from Kafka, Kinesis, or websocket connections.

You can implement a batching solution in Python with Psycopg2.
You can implement the ingestion logic within the `on_event` function that
you can then pass over to the websocket object.

This function needs to:

1.  Check if the item is a data item, and not websocket metadata.
1.  Adjust the data so that it fits the database schema, including the data
    types, and order of columns.
1.  Add it to the in-memory batch, which is a list in Python.
1.  If the batch reaches a certain size, insert the data, and reset or empty the list.

> ## Ingesting data in real-time

Reviewer: This procedure is still broken:

![image](https://github.com/timescale/docs/assets/3914967/20c85ca4-bc5b-4e8f-b6f4-2003d9e1a2f0)

    ORDER BY bucket;
    ```

    The result of the query looks like this:

    ```sql
             bucket         | symbol  |  open   |  high   |   low   |  close  | day_volume
    ------------------------+---------+---------+---------+---------+---------+------------
     2023-05-30 08:00:00+00 | AAPL   | 176.31 | 176.31 |    176 | 176.01 |           
     2023-05-30 08:01:00+00 | AAPL   | 176.27 | 176.27 | 176.02 |  176.2 |           
     2023-05-30 08:06:00+00 | AAPL   | 176.03 | 176.04 | 175.95 |    176 |           
     2023-05-30 08:07:00+00 | AAPL   | 175.95 |    176 | 175.82 | 175.91 |           
     2023-05-30 08:08:00+00 | AAPL   | 175.92 | 176.02 |  175.8 | 176.02 |           
     2023-05-30 08:09:00+00 | AAPL   | 176.02 | 176.02 |  175.9 | 175.98 |           
     2023-05-30 08:10:00+00 | AAPL   | 175.98 | 175.98 | 175.94 | 175.94 |           
     2023-05-30 08:11:00+00 | AAPL   | 175.94 | 175.94 | 175.91 | 175.91 |           
     2023-05-30 08:12:00+00 | AAPL   |  175.9 | 175.94 |  175.9 | 175.94 |
    ```

</Procedure>

</Collapsible>
    
<Collapsible heading="Visualize the OHLCV data in Grafana" defaultExpanded={false}>

>     You can vizualize the OHLCV data that you created using the queries in Grafana.

Reviewer: ```suggestion

    You can visualize the OHLCV data that you created using the queries in Grafana.
```

> <GraphOhlcv />

Reviewer: ```suggestion

<GraphOhlcv />
```
